metrics/prometheus.go

```diff
@@ -0,0 +1,183 @@
+// implements prometheus.Collector.
+func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
+	ch <- e.metrics.UpTime.Desc()
+	ch <- e.metrics.RequestResponseTimeTotal.Desc()
+	ch <- e.metrics.RequestResponseTimeAvg.Desc()
+
+	for _, metric := range e.metrics.RequestStatusCountCurrent {
+		ch <- metric.Desc()
+	}
+	for _, metric := range e.metrics.RequestStatusCountTotal {
+		ch <- metric.Desc()
+	}
+}
+
+// Collect delivers the Traefik stats to Prometheus
+// as Prometheus metrics. It implements prometheus.Collector.
+func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+	e.mutex.Lock() // To protect metrics from concurrent collects.
+	defer e.mutex.Unlock()
+
+	e.refresh()
+
+	ch <- e.metrics.UpTime
+	ch <- e.metrics.RequestResponseTimeTotal
+	ch <- e.metrics.RequestResponseTimeAvg
+
+	for _, metric := range e.metrics.RequestStatusCountCurrent {
+		ch <- metric
+	}
+	for _, metric := range e.metrics.RequestStatusCountTotal {
+		ch <- metric
+	}
+}
+
+func (e *Exporter) refresh() {
+	var data = Metrics.Data()
+
+	e.metrics.UpTime.Set(data.UpTimeSec)
+	e.metrics.RequestResponseTimeTotal.Set(data.TotalResponseTimeSec)
+	e.metrics.RequestResponseTimeAvg.Set(data.AverageResponseTimeSec)
+
+	// Current request count, labeled by statusCode
+	// Must be reset for missing status code metrics in data
+	for _, metric := range e.metrics.RequestStatusCountCurrent {
+		metric.Set(0)
+	}
+	for statusCode, nbr := range data.StatusCodeCount {
+		if _, ok := e.metrics.RequestStatusCountCurrent[statusCode]; ok == false {
+			e.metrics.RequestStatusCountCurrent[statusCode] = newGauge("request_count_current", "Number of request handled by Traefik", prometheus.Labels{"statusCode": statusCode})
+		}
+		e.metrics.RequestStatusCountCurrent[statusCode].Set(float64(nbr))
+	}
+
+	// Total request count, labeled by statusCode
+	for statusCode, nbr := range data.TotalStatusCodeCount {
+		if _, ok := e.metrics.RequestStatusCountTotal[statusCode]; ok == false {
+			e.metrics.RequestStatusCountTotal[statusCode] = newGauge("request_count_total", "Number of request handled by Traefik", prometheus.Labels{"statusCode": statusCode})
+		}
+		e.metrics.RequestStatusCountTotal[statusCode].Set(float64(nbr))
+	}
+}
+
+var exporter *Exporter
+
+func init() {
+	exporter = &Exporter{
+		metrics: globalMetrics{
+			UpTime: newGauge("uptime", "Current Traefik uptime", nil),
+			RequestResponseTimeTotal:  newGauge("request_response_time_total", "Total response time of Traefik requests", nil),
+			RequestResponseTimeAvg:    newGauge("request_response_time_avg", "Average response time of Traefik requests", nil),
+			RequestStatusCountCurrent: map[string]prometheus.Gauge{},
+			RequestStatusCountTotal:   map[string]prometheus.Gauge{},
+		},
+	}
+}
+
+// GetPrometheusHandler expose Metrics data in the Prometheus format on /metrics
+func GetPrometheusHandler() http.Handler {
+	prometheus.MustRegister(exporter)
+	prometheus.Unregister(prometheus.NewGoCollector())
```

Comment:
I used to disable go metrics in my others prometheus exporters, but is this case, the exporter is part of the binary, so we can activate it !